File: totoggle_go/internal/toggle/application.go
package toggle

// Application is the full set of toggles fetched for one app, indexed for O(1) path lookup. It
// owns the path-cascade logic (AncestorsOf) once, instead of each caller (cache, client) walking
// segments itself.
type Application struct {
	Toggles []Toggle
	byPath  map[string]Toggle
}

// NewApplication indexes toggles by path for lookup.
func NewApplication(toggles []Toggle) Application {
	byPath := make(map[string]Toggle, len(toggles))
	for _, tg := range toggles {
		byPath[tg.Path.String()] = tg
	}
	return Application{Toggles: toggles, byPath: byPath}
}

// ByPath looks up the toggle at an exact path.
func (a Application) ByPath(p Path) (Toggle, bool) {
	tg, ok := a.byPath[p.String()]
	return tg, ok
}

// AncestorsOf returns every toggle on the path from root to (but not including) p that is present
// in this Application, root first. An ancestor segment that was never configured is skipped
// rather than treated as an error — it simply doesn't contribute to the cascade.
func (a Application) AncestorsOf(p Path) []Toggle {
	ancestorPaths := p.AncestorPaths()
	ancestors := make([]Toggle, 0, len(ancestorPaths))
	for _, ap := range ancestorPaths {
		if tg, ok := a.byPath[ap.String()]; ok {
			ancestors = append(ancestors, tg)
		}
	}
	return ancestors
}

File: totoggle_node/src/internal/toggle/application.ts
import type { Path } from "./path.js";
import type { Toggle } from "./toggle.js";

/**
 * The full set of toggles fetched for one application, indexed for O(1) path lookup. Owns the
 * path-cascade logic (ancestorsOf) once, instead of each caller (cache, client) walking segments
 * itself.
 */
export class Application {
  readonly toggles: readonly Toggle[];
  private readonly byPathIndex: Map<string, Toggle>;

  constructor(toggles: readonly Toggle[]) {
    this.toggles = toggles;
    this.byPathIndex = new Map(toggles.map((t) => [t.path.toString(), t]));
  }

  /** Looks up the toggle at an exact path. */
  byPath(path: Path): Toggle | undefined {
    return this.byPathIndex.get(path.toString());
  }

  /** Every toggle on the path from root to (but not including) path that is present in this
   * Application, root first. An ancestor segment that was never fetched is skipped rather than
   * treated as an error. */
  ancestorsOf(path: Path): Toggle[] {
    const ancestors: Toggle[] = [];
    for (const ancestorPath of path.ancestorPaths()) {
      const toggle = this.byPathIndex.get(ancestorPath.toString());
      if (toggle) {
        ancestors.push(toggle);
      }
    }
    return ancestors;
  }
}
